### PR TITLE
Introduce GetPods() to replace GetKubeletDockerContainers()

### PR DESCRIPTION
**Pull request by yifan-gu**
A follow-up fix for #5526
To smooth the review, I only replaced in the syncPod/SyncPods. There are other places needed to be cleanup. Let's do it in following PRs.

PTAL thank you! @dchen1107 @vmarmol

### CODE REVIEW

**Comment by yifan-gu** [comment]
Tests fixed. Going to see if it can pass the e2e.

File: pkg/kubelet/container/runtime.go [hunk 1]
```diff
@@ -81,4 +81,34 @@ type Container struct {
...
+	for i := range p {
+		if p[i].ID == podUID {
+			return *p[i]
+		}
+	}
+	return Pod{}
+}
+
+// FindContainerByName returns a container in the pod with the given name.
+// Note that when there are multiple containers with the same name, then only
```
Comment (vmarmol) [inline pkg/kubelet/container/runtime.go]: nit: We can just say that the first match will be returned, the user can have the data sorted in whatever order they'd like.

File: pkg/kubelet/container/runtime.go [hunk 2]
```diff
@@ -81,4 +81,34 @@ type Container struct {
...
 	// not managed by kubelet.
 	Hash uint64
+	// The timestamp of the creation time of the container.
+	// TODO(yifan): Consider to move it to api.ContainerStatus.
+	Created int64
+}
+
+type Pods []*Pod
+
+// FindPodByID returns a pod in the pod list by UID. It returns nil
```
Comment (xiang90) [inline pkg/kubelet/container/runtime.go]: The comment does not match the function. It does not return nil when not found.

File: pkg/kubelet/container/runtime.go (hunk 1 above)
Comment (xiang90) [inline pkg/kubelet/container/runtime.go]: +1.

File: pkg/kubelet/container/runtime.go [hunk 3]
```diff
@@ -81,4 +81,34 @@ type Container struct {
...
+			return *p[i]
+		}
+	}
+	return Pod{}
+}
+
+// FindContainerByName returns a container in the pod with the given name.
+// Note that when there are multiple containers with the same name, then only
+// the newest container with the given name will be return (assuming containers
+// are sorted in reversed time order).
```
Comment (xiang90) [inline pkg/kubelet/container/runtime.go]: It returns nil if not found.

File: pkg/kubelet/dockertools/docker.go [hunk 4]
```diff
@@ -812,3 +813,65 @@ type ContainerCommandRunner interface {
 	ExecInContainer(containerID string, cmd []string, in io.Reader, out, err io.WriteCloser, tty bool) error
 	PortForward(podInfraContainerID string, port uint16, stream io.ReadWriteCloser) error
 }
+
+// Parse the pod full name. TODO(yifan): This is duplicated with kubelet.ParsePodFullName.
+func parsePodFullName(podFullName string) (string, string, error) {
```
Comment (vmarmol) [inline pkg/kubelet/dockertools/docker.go]: We can replace other uses of this to keep only 1 version.

File: pkg/kubelet/dockertools/docker.go [hunk 5]
```diff
@@ -812,3 +813,65 @@ type ContainerCommandRunner interface {
...
+			glog.Warningf("Cannog parse empty docker container name: %#v", c.Names)
+			continue
+		}
+		podFullName, podUID, containerName, hash, err := ParseDockerName(c.Names[0])
+		if err != nil {
+			glog.Warningf("Parse docker container name %q error: %v", c.Names[0], err)
+			continue
+		}
+
+		// TODO(yifan): Seems to break container VM users?
```
Comment (vmarmol) [inline pkg/kubelet/dockertools/docker.go]: Container VMs now have UID, no user upgrades Kubelets on those so a new Kubelet will come with new containers.

File: pkg/kubelet/dockertools/docker.go [hunk 6]
```diff
@@ -812,3 +813,65 @@ type ContainerCommandRunner interface {
...
+		// https://github.com/GoogleCloudPlatform/kubernetes/pull/5032#issuecomment-77432277
+		pod, found := pods[podUID]
+		if !found {
+			name, namespace, err := parsePodFullName(podFullName)
+			if err != nil {
+				// TODO(yifan): Or skip the pod?
```
Comment (vmarmol) [inline pkg/kubelet/dockertools/docker.go]: Our current policy is to skip

File: pkg/kubelet/container/runtime.go (hunk 3 above)
Comment (yifan-gu) [inline pkg/kubelet/container/runtime.go]: :(

File: pkg/kubelet/dockertools/docker.go (hunk 4 above)
Comment (yifan-gu) [inline pkg/kubelet/dockertools/docker.go]: Ye, it causes circular imports. How about putting it in the new runtime.go? @vmarmol

Comment (vmarmol) [inline pkg/kubelet/dockertools/docker.go]: Yes I think so too :) the parsing and unparsing of the podFullName should only happen in one place.

**Comment by yifan-gu** [comment]
@vmarmol I addressed the comments. Moved GodPodFullname and others into the pkg/kubelet/container package.
I think the build is failing because of the master. Waiting for the master to be healthy

**Comment by yifan-gu** [comment]
e2e fails 3 tests. I have the same result with HEAD fe65aa5dc98268f0a1a311faa63a272a4e1f16e1 


```
Summarizing 3 Failures:

[Fail] PD [It] should schedule a pod w/ a readonly PD on two hosts, then remove both.
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:154

[Fail] Cluster level logging using Elasticsearch [It] should check that logs from pods on all nodes are ingested into Elasticsearch
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:337

[Fail] Services [It] should correctly serve identically named services in different namespaces on different external IP addresses
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:372

Ran 29 of 31 Specs in 2204.132 seconds
FAIL! -- 26 Passed | 3 Failed | 2 Pending | 0 Skipped watchEvents output: |

File: pkg/kubelet/kubelet.go [hunk 7]
```diff
@@ -1580,31 +1584,27 @@ func (kl *Kubelet) SyncPods(allPods []api.Pod, podSyncTypes map[types.UID]metric
 
 	// Kill any containers we don't need.
 	killed := []string{}
-	for ix := range dockerContainers {
-		// Don't kill containers that are in the desired pods.
-		dockerName, _, err := dockertools.ParseDockerName(dockerContainers[ix].Names[0])
-		_, found := desiredPods[dockerName.PodUID]
-		if err == nil && found {
+	for _, pod := range runningPods {
```
Comment (dchen1107) [inline pkg/kubelet/kubelet.go]: We might misread the logic here, but shouldn't you introduce a bug in a update case described as the below:

 A running pod with two containers: c1 and c2, now update pod with only c1. With your new change, c2 couldn't be removed?

Comment (yifan-gu) [inline pkg/kubelet/kubelet.go]: @dchen1107 I think the syncPod() will take care of the update case.

So the logic I want to express here is:

```
foreach desired pod:
    find its corresponding running pod, and do syncPod.
    
foreach running pod:
    if it is not desired, kill the whole pod.
```

The first loop makes sure we have all the pod up to the desired state. The second loop makes sure we remove any unnecessary pods.

The assumption here is all containers we created can be grouped into a pod.

Comment (dchen1107) [inline pkg/kubelet/kubelet.go]: ACK.

**Comment by dchen1107** [comment]
LGTM. Could you please rebase again for merge?

**Comment by yifan-gu** [comment]
Rebased. Running the e2e again.

**Comment by vmarmol** [comment]
LGTM @yifan-gu thanks!

**Comment by yifan-gu** [comment]
e2e result:
```
Summarizing 1 Failure:

[Fail] Services [It] should correctly serve identically named services in different namespaces on different external IP addresses
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:372

Ran 29 of 31 Specs in 1447.767 seconds
FAIL! -- 28 Passed | 1 Failed | 2 Pending | 0 Skipped watchEvents output: |
```
I think this is not related with this PR: #5722

**Comment by yifan-gu** [comment]
...shippable is failing me again :(

**Comment by dchen1107** [comment]
I kicked out shippable again. we can merge once it is green.

**Comment by yifan-gu** [comment]
@dchen1107 Thank you! Since I don't have access to the shippable's host of kubenetes repo, so I am not sure about the reason.
But my personal fork also fails on the master e1a66ad01569006615f6da2aff7c30e5367bfdce.

```
 GOPATH=$PWD/Godeps/_workspace:$GOPATH go install ./...
cluster/addons/dns/kube2sky/kube2sky.go:30:2: cannot find package "github.com/GoogleCloudPlatform/kubernetes/pkg/api" in any of:
	/home/shippable/.gvm/gos/go1.4/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api (from $GOROOT)
	/home/shippable/workspace/src/github.com/yifan-gu/kubernetes/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api (from $GOPATH)
	/home/shippable/workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api
cluster/addons/dns/kube2sky/kube2sky.go:31:2: cannot find package "github.com/GoogleCloudPlatform/kubernetes/pkg/client" in any of:
	/home/shippable/.gvm/gos/go1.4/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client (from $GOROOT)
	/home/shippable/workspace/src/github.com/yifan-gu/kubernetes/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client (from $GOPATH)
	/home/shippable/workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client
cluster/addons/dns/kube2sky/kube2sky.go:32:2: cannot find package "github.com/GoogleCloudPlatform/kubernetes/pkg/fields" in any of:
	/home/shippable/.gvm/gos/go1.4/src/github.com/GoogleCloudPlatform/kubernetes/pkg/fields (from $GOROOT)
	/home/shippable/workspace/src/github.com/yifan-gu/kubernetes/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/fields (from $GOPATH)
	/home/shippable/workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/fields

...

```

**Comment by dchen1107** [comment]
All tests are passed, but unfortunately it looks like you need to rebase again. Sorry about such churns, but good part is that kubelet side changed dramatically everyday. :-)

**Comment by yifan-gu** [comment]
@dchen1107 No worries. Just rebased. Wait for green :)

**Comment by yifan-gu** [comment]
e2e looks good, only the same test fails.

**Comment by dchen1107** [comment]
Merge now since shippable is green and that e2e test failed on HEAD to me today.

**Comment by yifan-gu** [comment]
@dchen1107 Thank you! Next I will try to figure out how to refactor the syncPod.

**Pull request merged**